Support GCC 7

Building Pony on Debian Sid works fine. When I try to build Pony on
Fedora Rawhide, I get:
```
Linking libponyrt-pic
main.c
Linking ponyc
build/release/libponyrt.a(pool.o): In function `pool_get.constprop.5':
pool.c:(.text+0x56b): undefined reference to `__atomic_compare_exchange_16'
build/release/libponyrt.a(pool.o): In function `ponyint_pool_alloc':
pool.c:(.text+0x6eb): undefined reference to `__atomic_compare_exchange_16'
build/release/libponyrt.a(pool.o): In function `ponyint_pool_free':
pool.c:(.text+0x8a3): undefined reference to `__atomic_compare_exchange_16'
build/release/libponyrt.a(pool.o): In function `ponyint_pool_alloc_size':
pool.c:(.text+0xc5f): undefined reference to `__atomic_compare_exchange_16'
build/release/libponyrt.a(pool.o): In function `ponyint_pool_free_size':
pool.c:(.text+0x1065): undefined reference to `__atomic_compare_exchange_16'
build/release/libponyrt.a(pool.o):pool.c:(.text+0x1132): more undefined references to `__atomic_compare_exchange_16' follow
collect2: error: ld returned 1 exit status
make: *** [Makefile:662: build/release/ponyc] Error 1
```
everything before this is fine. I am guessing this is a GCC version
thing. Debian Sid has GCC 6.3 where Fedora Rawhide has GCC 7.0. 

Update 2017-08-05: This is now a problem on Debian Sid as it has just switched to GCC 7 as the default compiler suite.

Just the end of the output, confirming the hypothesis.:
```
cc -MMD -MP -march=native -mtune=generic -Werror -Wconversion -Wno-sign-conversion -Wextra -Wall -mcx16 -O3 -DNDEBUG -std=gnu11 -fexceptions -DPONY_VERSION=\"0.11.4-077c27605\" -DLLVM_VERSION=\"3.9.1\" -DPONY_COMPILER=\"cc\" -DPONY_ARCH=\"native\" -DBUILD_COMPILER=\""cc (GCC) 7.0.1 20170309 (Red Hat 7.0.1-0.12)"\" -DPONY_BUILD_CONFIG=\"release\" -D_FILE_OFFSET_BITS=64  -fpic -c -o build/release/obj/libponyrt-pic/gc/cycle.o src/libponyrt/gc/cycle.c  -I src/common/ -I src/libponyrt/
gc.c
cc -MMD -MP -march=native -mtune=generic -Werror -Wconversion -Wno-sign-conversion -Wextra -Wall -mcx16 -O3 -DNDEBUG -std=gnu11 -fexceptions -DPONY_VERSION=\"0.11.4-077c27605\" -DLLVM_VERSION=\"3.9.1\" -DPONY_COMPILER=\"cc\" -DPONY_ARCH=\"native\" -DBUILD_COMPILER=\""cc (GCC) 7.0.1 20170309 (Red Hat 7.0.1-0.12)"\" -DPONY_BUILD_CONFIG=\"release\" -D_FILE_OFFSET_BITS=64  -fpic -c -o build/release/obj/libponyrt-pic/gc/gc.o src/libponyrt/gc/gc.c  -I src/common/ -I src/libponyrt/
Linking libponyrt-pic
gcc-ar rcs build/release/libponyrt-pic.a build/release/obj/libponyrt-pic/actor/actor.o build/release/obj/libponyrt-pic/actor/messageq.o build/release/obj/libponyrt-pic/ds/fun.o build/release/obj/libponyrt-pic/ds/hash.o build/release/obj/libponyrt-pic/ds/stack.o build/release/obj/libponyrt-pic/ds/list.o build/release/obj/libponyrt-pic/options/options.o build/release/obj/libponyrt-pic/mem/pool.o build/release/obj/libponyrt-pic/mem/heap.o build/release/obj/libponyrt-pic/mem/pagemap.o build/release/obj/libponyrt-pic/mem/alloc.o build/release/obj/libponyrt-pic/asio/event.o build/release/obj/libponyrt-pic/asio/epoll.o build/release/obj/libponyrt-pic/asio/asio.o build/release/obj/libponyrt-pic/platform/threads.o build/release/obj/libponyrt-pic/platform/ponyassert.o build/release/obj/libponyrt-pic/lang/lsda.o build/release/obj/libponyrt-pic/lang/posix_except.o build/release/obj/libponyrt-pic/lang/io.o build/release/obj/libponyrt-pic/lang/time.o build/release/obj/libponyrt-pic/lang/stdfd.o build/release/obj/libponyrt-pic/lang/socket.o build/release/obj/libponyrt-pic/lang/paths.o build/release/obj/libponyrt-pic/lang/directory.o build/release/obj/libponyrt-pic/lang/ssl.o build/release/obj/libponyrt-pic/lang/stat.o build/release/obj/libponyrt-pic/sched/mpmcq.o build/release/obj/libponyrt-pic/sched/scheduler.o build/release/obj/libponyrt-pic/sched/cpu.o build/release/obj/libponyrt-pic/sched/start.o build/release/obj/libponyrt-pic/gc/trace.o build/release/obj/libponyrt-pic/gc/actormap.o build/release/obj/libponyrt-pic/gc/serialise.o build/release/obj/libponyrt-pic/gc/objectmap.o build/release/obj/libponyrt-pic/gc/delta.o build/release/obj/libponyrt-pic/gc/cycle.o build/release/obj/libponyrt-pic/gc/gc.o
gtest-all.cc
g++ -MMD -MP -march=native -mtune=generic -Werror -Wall -mcx16 -O3 -DNDEBUG -std=gnu++11 -fno-rtti   -c -o build/release/obj/libgtest/gtest-all.o lib/gtest/gtest-all.cc  -isystem lib/gtest/
Linking libgtest
gcc-ar rcs build/release/libgtest.a build/release/obj/libgtest/gtest-all.o
gbenchmark_main.cc
g++ -MMD -MP -march=native -mtune=generic -Werror -Wall -mcx16 -O3 -DNDEBUG -std=gnu++11 -fno-rtti  -DHAVE_POSIX_REGEX -c -o build/release/obj/libgbenchmark/gbenchmark_main.o lib/gbenchmark/gbenchmark_main.cc  -isystem lib/gbenchmark/include/
gbenchmark-all.cc
g++ -MMD -MP -march=native -mtune=generic -Werror -Wall -mcx16 -O3 -DNDEBUG -std=gnu++11 -fno-rtti  -DHAVE_POSIX_REGEX -c -o build/release/obj/libgbenchmark/gbenchmark-all.o lib/gbenchmark/gbenchmark-all.cc  -isystem lib/gbenchmark/include/
Linking libgbenchmark
gcc-ar rcs build/release/libgbenchmark.a build/release/obj/libgbenchmark/gbenchmark_main.o build/release/obj/libgbenchmark/gbenchmark-all.o
main.c
cc -MMD -MP -march=native -mtune=generic -Werror -Wconversion -Wno-sign-conversion -Wextra -Wall -mcx16 -O3 -DNDEBUG -std=gnu11 -fexceptions -DPONY_VERSION=\"0.11.4-077c27605\" -DLLVM_VERSION=\"3.9.1\" -DPONY_COMPILER=\"cc\" -DPONY_ARCH=\"native\" -DBUILD_COMPILER=\""cc (GCC) 7.0.1 20170309 (Red Hat 7.0.1-0.12)"\" -DPONY_BUILD_CONFIG=\"release\" -D_FILE_OFFSET_BITS=64  -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -c -o build/release/obj/ponyc/main.o src/ponyc/main.c  -I src/common/ -I src/libponyrt/ 
Linking ponyc
g++  -o build/release/ponyc build/release/obj/ponyc/main.o -march=native -mtune=generic -mcx16 -L build/release -L/usr/lib64  -lponyc -lponyrt -lLLVM-3.9 -lz -lncurses -lpthread -ldl -rdynamic
build/release/libponyrt.a(pool.o): In function `pool_get.constprop.5':
pool.c:(.text+0x56b): undefined reference to `__atomic_compare_exchange_16'
build/release/libponyrt.a(pool.o): In function `ponyint_pool_alloc':
pool.c:(.text+0x6eb): undefined reference to `__atomic_compare_exchange_16'
build/release/libponyrt.a(pool.o): In function `ponyint_pool_free':
pool.c:(.text+0x8a3): undefined reference to `__atomic_compare_exchange_16'
build/release/libponyrt.a(pool.o): In function `ponyint_pool_alloc_size':
pool.c:(.text+0xc5f): undefined reference to `__atomic_compare_exchange_16'
build/release/libponyrt.a(pool.o): In function `ponyint_pool_free_size':
pool.c:(.text+0x1065): undefined reference to `__atomic_compare_exchange_16'
build/release/libponyrt.a(pool.o):pool.c:(.text+0x1132): more undefined references to `__atomic_compare_exchange_16' follow
collect2: error: ld returned 1 exit status
make: *** [Makefile:662: build/release/ponyc] Error 1
```

The Fedora experiment on a compiled up repository:
```
|> make verbose=1
Linking ponyc
g++  -o build/release/ponyc build/release/obj/ponyc/main.o -march=native -mtune=generic -mcx16 -L build/release -L/usr/lib64  -lponyc -lponyrt -lLLVM-3.9 -lz -lncurses -lpthread -ldl -rdynamic
build/release/libponyrt.a(pool.o): In function `pool_get.constprop.5':
pool.c:(.text+0x56b): undefined reference to `__atomic_compare_exchange_16'
build/release/libponyrt.a(pool.o): In function `ponyint_pool_alloc':
pool.c:(.text+0x6eb): undefined reference to `__atomic_compare_exchange_16'
build/release/libponyrt.a(pool.o): In function `ponyint_pool_free':
pool.c:(.text+0x8a3): undefined reference to `__atomic_compare_exchange_16'
build/release/libponyrt.a(pool.o): In function `ponyint_pool_alloc_size':
pool.c:(.text+0xc5f): undefined reference to `__atomic_compare_exchange_16'
build/release/libponyrt.a(pool.o): In function `ponyint_pool_free_size':
pool.c:(.text+0x1065): undefined reference to `__atomic_compare_exchange_16'
build/release/libponyrt.a(pool.o):pool.c:(.text+0x1132): more undefined references to `__atomic_compare_exchange_16' follow
collect2: error: ld returned 1 exit status
make: *** [Makefile:662: build/release/ponyc] Error 1
1011 anglides:/home/Checkouts/Git/PonyC (git:master)
|> g++  -o build/release/ponyc build/release/obj/ponyc/main.o -march=native -mtune=generic -mcx16 -L build/release -L/usr/lib64  -lponyc -lponyrt -lLLVM-3.9 -lz -lncurses -lpthread -latomic -ldl -rdynamic
/usr/bin/ld: cannot find /usr/lib64/libatomic.so.1.2.0
collect2: error: ld returned 1 exit status
```

I have now manually install libatomic on Fedora, but this is now a different computer, a laptop rather than a workstation. This is though a good test since it is a different processor and memory architecture. Just running "make verbose=1" on the Fedora Rawhide laptop results in the same behaviour as before even with libatomic installed:
```
Linking ponyc
g++  -o build/release/ponyc build/release/obj/ponyc/main.o -march=native -mtune=generic -mcx16 -L build/release -L/usr/lib64  -lponyc -lponyrt -lLLVM-3.9 -lz -lncurses -lpthread -ldl -rdynamic
build/release/libponyrt.a(pool.o): In function `pool_get.constprop.5':
pool.c:(.text+0x56b): undefined reference to `__atomic_compare_exchange_16'
build/release/libponyrt.a(pool.o): In function `ponyint_pool_alloc':
pool.c:(.text+0x6eb): undefined reference to `__atomic_compare_exchange_16'
build/release/libponyrt.a(pool.o): In function `ponyint_pool_free':
pool.c:(.text+0x8a3): undefined reference to `__atomic_compare_exchange_16'
build/release/libponyrt.a(pool.o): In function `ponyint_pool_alloc_size':
pool.c:(.text+0xc5f): undefined reference to `__atomic_compare_exchange_16'
build/release/libponyrt.a(pool.o): In function `ponyint_pool_free_size':
pool.c:(.text+0x1065): undefined reference to `__atomic_compare_exchange_16'
build/release/libponyrt.a(pool.o):pool.c:(.text+0x1132): more undefined references to `__atomic_compare_exchange_16' follow
collect2: error: ld returned 1 exit status
make: *** [Makefile:662: build/release/ponyc] Error 1
1006 lavaine:/home/Checkouts/Git/PonyC (git:master)
|> g++  -o build/release/ponyc build/release/obj/ponyc/main.o -march=native -mtune=generic -mcx16 -L build/release -L/usr/lib64  -lponyc -lponyrt -lLLVM-3.9 -latomic -lz -lncurses -lpthread -ldl -rdynamic
```
so whilst an explicit -latomic is needed on Fedora it is not on Debian. I wonder if using CMake, SCons or Meson, would get rid of the problem as the use of pkg-config should get all the dependencies right. Of course this can be done in Make, but it is a lot harder.

The real concern though, given the separateness of the atomics library, how did this ever work on Debian?

I'd like to suggest raising the priority on this one, if I may. Whilst this used to be a Rawhide only problem, it is now a Debian Sid problem: Debian Sid has just moved from GCC 6 to GCC 7 as the default compiler suite, so the bug originally reported as not being a problem on Debian Sid is now a problem, on Debian Sid:
```
build/release/libponyrt.a(pool.o): In function `pool_get.constprop.5':
pool.c:(.text+0x4aa): undefined reference to `__atomic_compare_exchange_16'
build/release/libponyrt.a(pool.o): In function `ponyint_pool_alloc':
pool.c:(.text+0x5da): undefined reference to `__atomic_compare_exchange_16'
build/release/libponyrt.a(pool.o): In function `ponyint_pool_free':
pool.c:(.text+0x73e): undefined reference to `__atomic_compare_exchange_16'
build/release/libponyrt.a(pool.o): In function `ponyint_pool_alloc_size':
pool.c:(.text+0x871): undefined reference to `__atomic_compare_exchange_16'
build/release/libponyrt.a(pool.o): In function `ponyint_pool_free_size':
pool.c:(.text+0xaee): undefined reference to `__atomic_compare_exchange_16'
build/release/libponyrt.a(pool.o):pool.c:(.text+0xdee): more undefined references to `__atomic_compare_exchange_16' follow
collect2: error: ld returned 1 exit status
Makefile:704: recipe for target 'build/release/ponyc' failed
make: *** [build/release/ponyc] Error 1
```

--- a/Makefile
+++ b/Makefile
@@ -463,11 +463,11 @@ libponyc.benchmarks.links = libgbenchmark libponyc libponyrt llvm
 libponyrt.benchmarks.links = libgbenchmark libponyrt
 
 ifeq ($(OSTYPE),linux)
-  ponyc.links += libpthread libdl
-  libponyc.tests.links += libpthread libdl
-  libponyrt.tests.links += libpthread libdl
-  libponyc.benchmarks.links += libpthread libdl
-  libponyrt.benchmarks.links += libpthread libdl
+  ponyc.links += libpthread libdl libatomic
+  libponyc.tests.links += libpthread libdl libatomic
+  libponyrt.tests.links += libpthread libdl libatomic
+  libponyc.benchmarks.links += libpthread libdl libatomic
+  libponyrt.benchmarks.links += libpthread libdl libatomic
 endif
 
 ifeq ($(OSTYPE),freebsd)

--- a/src/libponyc/codegen/genexe.c
+++ b/src/libponyc/codegen/genexe.c
@@ -272,6 +272,7 @@ static bool link_exe(compile_t* c, ast_t* program,
   const char* export = target_is_linux(c->opt->triple) ?
     "-Wl,--export-dynamic-symbol=__PonyDescTablePtr "
     "-Wl,--export-dynamic-symbol=__PonyDescTableSize" : "-rdynamic";
+  const char* atomic = target_is_linux(c->opt->triple) ? "-latomic" : "";
 
   size_t ld_len = 512 + strlen(file_exe) + strlen(file_o) + strlen(lib_args)
                   + strlen(arch) + strlen(mcx16_arg) + strlen(fuseld)
@@ -284,9 +285,9 @@ static bool link_exe(compile_t* c, ast_t* program,
 #ifdef PONY_USE_LTO
     "-flto -fuse-linker-plugin "
 #endif
-    "%s %s %s %s -lpthread %s -lm %s",
-    linker, file_exe, arch, mcx16_arg, fuseld, file_o, lib_args, ponyrt, ldl,
-    export
+    "%s %s %s %s -lpthread %s %s -lm %s",
+    linker, file_exe, arch, mcx16_arg, atomic, fuseld, file_o, lib_args,
+    ponyrt, ldl, export
     );
 
   if(c->opt->verbosity >= VERBOSITY_TOOL_INFO)

